src/main/java/hudson/remoting/Engine.java

```diff
@@ -758,11 +759,18 @@ private void innerRun(IOHub hub, SSLContext context, ExecutorService service) {
                 final JnlpAgentEndpoint endpoint;
                 try {
                     endpoint = resolver.resolve();
-                } catch (Exception e) {
-                    if (Boolean.getBoolean(Engine.class.getName() + ".nonFatalJnlpAgentEndpointResolutionExceptions")) {
-                        events.status("Could not resolve JNLP agent endpoint", e);
+                } catch (IOException e) {
+                    if (!noReconnect) {
+                        events.status("Could not locate server among " + candidateUrls + "; waiting 10 seconds before retry", e);
+                        // TODO refactor various sleep statements into a common method
+                        TimeUnit.SECONDS.sleep(10);
+                        continue;
                     } else {
-                        events.error(e);
+                        if (Boolean.getBoolean(Engine.class.getName() + ".nonFatalJnlpAgentEndpointResolutionExceptions")) {
+                            events.status("Could not resolve JNLP agent endpoint", e);
```

Comment:
Note to readers excluding the author (not a review): this property derives from #449, where `SwarmClient` wraps `hudson.remoting.jnlp.Main` in-JVM and thus the call to `System.exit` would be “fatal”. For normal uses of Remoting, the distinction is merely between logging a stack trace at `SEVERE` and exiting with -1 vs. logging a stack trace at `INFO` and exiting with 0 (but still effectively treating the error as fatal).

Reply:
> thus the call to System.exit would be “fatal”

Not sure why scare quotes were used here, but yes a call to `System.exit` is fatal to the life of the process.